### PR TITLE
2021 01 25 type maximalist

### PR DESCRIPTION
**Pull request by Christewart**
Add some types for things inside of the CETCalculator like `CETOutcome` `MultiOracleDigits`, `MultiOracleOutcome`.

### CODE REVIEW

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala [hunk 1]
```diff
@@ -369,23 +392,34 @@ object CETCalculator {
...
 
     ranges.flatMap { range =>
       range match {
         case StartZero(indexFrom, indexTo) =>
-          groupByIgnoringDigits(indexFrom, indexTo, base, numDigits).map(
-            _ -> Satoshis.zero)
+          groupByIgnoringDigits(indexFrom, indexTo, base, numDigits).map {
+            decomp =>
+              val payout = Satoshis.zero
+              CETOutcome(decomp, payout)
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala]: nit: On all of these I prefer the one liner `CETOutcome(decomp, payout = Satoshis.zero)`

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala [hunk 2]
```diff
@@ -13,6 +13,30 @@ import scala.annotation.tailrec
...
+    *
+    * or if we are using base 16 (hex)
+    * Vector(9,8,10,11,2,3,2,4,1,15)
+    */
+  type Digits = Vector[Int]
+
+  /** A Vector of Digits corresponding to each oracle in order */
+  case class MultiOracleDigits(vec: Vector[Digits]) {
+    def toVector: Vector[Digits] = vec
+  }
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala]: I think I prefer this as a type alias after having looked at the code but everything else looks good imo

**Review by nkohen — COMMENTED** [review]
`MulitOracleDigits` is my only constraint

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala (hunk 2 above)
Comment (Christewart) [reply core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala]: If you are indifferent I would really like to keep this. I find it frustrating to work with type aliases in part due to the scala compiler errors. With type aliases on type aliases (`Digits` inside of `MultiOracleDigits`) you can get error messages that look like this 

```scala
[error]  required: org.bitcoins.core.protocol.dlc.CETCalculator.MultiOracleDigits
[error]     (which expands to)  scala.collection.immutable.Vector[scala.collection.immutable.Vector[Int]]
[error]         builder.++=(Vector(1))
```

as compared to what I have on c15cbde5478c1b14fa932353ea886d8e6bb6cc64

```scala
[error] /home/chris/dev/bitcoin-s/core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala:718:28: type mismatch;
[error]  found   : scala.collection.immutable.Vector[Int]
[error]  required: org.bitcoins.core.protocol.dlc.CETCalculator.MultiOracleDigits
[error]           builder.+=(Vector(1))
```

I find it hard to reason about type aliases inside of type aliases, but maybe this is just a personal preference. If you really want me to switch it back I will.

**Comment by Christewart** [comment]
I am seeing some failures on CI unrelated to this PR (#2455 will get included on `adaptor-dlc` on the next rebase).

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala (hunk 2 above)
Comment (nkohen) [reply core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala]: Hm, between those two error messages I find the first much more informative? It is literally the second one but with even more extra information?

Comment (Christewart) [reply core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala]: Yes, that's the problem. I find it to be a wall of text that includes unnecessary information if I'm just trying to get things to compile (the type I care about is `MultiOracleDigits`, not trying to grok that the implementation of `MultiOracleDigits` is a `Vector[Vector[Int]]`.

This case isn't really that bad, but if you go hog wild with type aliases like the scala slick project or previous iterations of suredbits-api, I find type aliases get expanded into walls of text by the scala compiler that are really hard to understand what is going on and maintain. Maybe scala just has a poor implementation of them, but it's the tool we are using

Any way, this case seems to be confined and we aren't going to agree so I will change the code for the sake of productivity.

File: app/gui/src/main/scala/org/bitcoins/gui/dlc/DLCPlotUtil.scala [hunk 3]
```diff
@@ -96,48 +97,45 @@ object DLCPlotUtil {
...
+      cets: Vector[CETOutcome],
       base: Int,
       numDigits: Int,
       executedCET: Vector[Int]): Figure = {
     plotCETs(cets, base, numDigits, executedCETOpt = Some(executedCET))
   }
 
   private def plotCETs(
-      cets: Vector[(Vector[Int], Satoshis)],
+      cets: Vector[CETOutcome], //(Vector[Int], Satoshis)
```
Comment (nkohen) [inline app/gui/src/main/scala/org/bitcoins/gui/dlc/DLCPlotUtil.scala]: ping

File: core-test/src/test/scala/org/bitcoins/core/protocol/dlc/CETCalculatorTest.scala [hunk 4]
```diff
@@ -334,11 +335,11 @@ class CETCalculatorTest extends BitcoinSUnitTest {
...
 
-    assert(coveringCETsMax.forall(_.length == 2))
-    assert(coveringCETsMin.forall(_.length == 2))
+    assert(coveringCETsMax.forall(_.toVector.length == 2))
+    assert(coveringCETsMin.forall(_.toVector.length == 2))
 
-    (coveringCETsMax.map(ds => (ds.head, ds.last)),
-     coveringCETsMin.map(ds => (ds.head, ds.last)))
+    (coveringCETsMax.map(ds => (ds.toVector.head, ds.toVector.last)),
+     coveringCETsMin.map(ds => (ds.toVector.head, ds.toVector.last)))
```
Comment (nkohen) [inline core-test/src/test/scala/org/bitcoins/core/protocol/dlc/CETCalculatorTest.scala]: I think these were missed in reverting?

File: core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala [hunk 5]
```diff
@@ -669,10 +707,13 @@ object CETCalculator {
           numToVec(start - minFail, numDigits, minFailExp)
         }
 
-        builder.++=(
+        val doubleCovering: Vector[Vector[Digits]] =
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/protocol/dlc/CETCalculator.scala]: nit: `Vector[MultiOracleDigits]`

File: core-test/src/test/scala/org/bitcoins/core/protocol/dlc/CETCalculatorTest.scala [hunk 6]
```diff
@@ -334,8 +335,8 @@ class CETCalculatorTest extends BitcoinSUnitTest {
                                               maximizeCoverage = false,
                                               numOracles = 2)
 
-    assert(coveringCETsMax.forall(_.length == 2))
-    assert(coveringCETsMin.forall(_.length == 2))
+    assert(coveringCETsMax.forall(_.toVector.length == 2))
+    assert(coveringCETsMin.forall(_.toVector.length == 2))
```
Comment (nkohen) [inline core-test/src/test/scala/org/bitcoins/core/protocol/dlc/CETCalculatorTest.scala]: I don't think these two calls are needed

**Review by nkohen — APPROVED** [review]
ACK

**Pull request merged**